### PR TITLE
perf(lua): don't copy tables and reduce for loops

### PR DESCRIPTION
**Pull request by futsuuu**
"reduce for loop" means the internal use of `vim.list_slice` and `vim.deep_equal`.

### CODE REVIEW

**Comment by tani** [comment]
Thank you very much for the outstanding optimization. I generally agree with it.  
I think the algorithm is sound, and there should be no problem merging it.

However, it seems that you are maintaining the state of `dmacro` with a global variable in the module, likely out of concern for the overhead of communication between Vim and Lua, right?

I have a sort of phobia of using global variables, so I’d like to avoid that if possible. Still, based on your experience, do you feel that the overhead introduced by changing this setup would be sufficiently painful? Could you share your thoughts, even just anecdotally?

**Comment by futsuuu** [comment]
I ran simple benchmarks:

```
len = 0
    vim.b:    0.014912ms
    table:    0.002164ms
len = 10 ^ 5
    vim.b:   41.518444ms
    table:    0.001901ms
len = 10 ^ 6
    vim.b:  265.648824ms
    table:    0.001696ms
len = 10 ^ 7
    vim.b: 2543.316751ms
    table:    0.001733ms
```

```lua
local len = 10 ^ 6
local list = {}
for _ = 1, len do
    table.insert(list, 'a')
end

local tbl = vim.b   -- `{}` vs `vim.b`
bench(function()
    tbl.foo = list
    list = tbl.foo
end)
```

The Lua table is always faster than `vim.b`, but in most cases we will quit Neovim before hitting the key 10 ^ 7 times (so, answer for your question "do you feel that the overhead introduced by changing this setup would be sufficiently painful?" is probably "No").

However, also `vim.b` can easily get/set values of other buffers.

```lua
local keys = vim.b[bufnr].dmacro_keys
local macro = vim.b[bufnr].dmacro_macro

local keys = buf_states[bufnr].keys
local macro = buf_states[bufnr].macro
```

So I want to use the Lua table for performance, even if it's just a difference we can't feel.
But if you still prefer `vim.b`, I'll remove a last commit.

**Comment by futsuuu** [comment]
I found more efficient optimizations so I changed this PR to draft, but it would be helpful if you could respond to the above.

**Comment by tani** [comment]
Thank you for the accurate feedback based on the benchmarks. I’m surprised that there can be such a noticeable difference in speed. However, as you pointed out, I now understand that it may not be something one would perceive directly in practice.

Having looked over your code, I’ve changed my mind and now agree with consolidating the code in Lua—in other words, I support not using vim.b. This is based on my intention to separate the Lua portion from the Vim portion as much as possible.

On the other hand, I still have reservations about using global variables. This is not a homework assignment for you, but rather my own future plan: as each buffer is created, I’m considering defining various functions and variables as closures (local functions), so that instead of sharing a single variable across all of Neovim, each buffer would have its own set of variables managed in Lua.

```luas
local buf_states =  { ... }
local function 

-- |
-- v

local function buf_new_hook()
  local buf_state = ...
  local function
end

-- autocmd for buf_new_hook
```

Once again, I am both interested in and grateful for your optimized code. The above is not a request, but a concept I have in mind. Still, if the idea I’ve mentioned could be incorporated into your optimized code, I would be even more pleased.

**Comment by futsuuu** [comment]
I'll create a new PR to stop managing multiple states in a variable.

File: lua/dmacro.lua [hunk 1]
```diff
@@ -23,100 +46,114 @@ end
...
     _M.set_state(keys, macro)
   end
 end
 
+--- Checks if the `keys` has a suffix that matches the previously used macro.
+--- @param macro dmacro.Range
+--- @param keys string[]
+--- @return boolean
+function _M.is_macro_still_valid(macro, keys)
+  return #keys == macro[2] or _M.span_equal(keys, macro[1], macro[2] + 1, macro[2] - macro[1] + 1)
```
Comment (tani) [inline lua/dmacro.lua]:
```suggestion
  local is_end_just_now = #keys == macro[2]
  local first_macro_start = macro[1]
  local second_macro_start = macro[2] + 1
  local macro_len = macro[2] - macro[1] + 1
  return is_end_just_now or M.span.equal(keys, first_macro_start, second_macro_start, macro_len)
```

File: lua/dmacro.lua [hunk 2]
```diff
@@ -23,100 +46,114 @@ end
...
     end
     _M.set_state(keys, macro)
   end
 end
 
+--- Checks if the `keys` has a suffix that matches the previously used macro.
+--- @param macro dmacro.Range
+--- @param keys string[]
+--- @return boolean
+function _M.is_macro_still_valid(macro, keys)
```
Comment (tani) [inline lua/dmacro.lua]: I prefer `has_prev_macro_suffix` instead of the name.

**Review by tani — CHANGES_REQUESTED** [review]
Thanks!

File: lua/dmacro.lua (hunk 1 above)
Comment (futsuuu) [reply lua/dmacro.lua]: Sorry, I noticed that this implementation doesn't check for a suffix...

**Review by tani — APPROVED** [review]
Thank you. Perfect!

**Pull request merged**